### PR TITLE
[CUDA graphs] Make stream semantics of backward calls consistent with other cuda ops

### PR DESCRIPTION
**Pull request by mcarilli**
Closed for resubmit to ci-all https://github.com/pytorch/pytorch/pull/57833
____________________

Right now, the autograd engine syncs backward leaf streams with the default stream, which makes the following pattern safe:
```python
with torch.cuda.stream(s):
    # imagine forward used many streams, so backward leaf nodes may run on many streams
    loss.backward()
# no sync
use grads
```
Comments in engine.cpp indicate this is for BC, but I don't think it's desirable.

First (as in the above code) the stream semantics of a backward() (or autograd.grad()) call relative to surrounding ops are inconsistent with the stream semantics of other calls.

Because of the inconsistency, in some cases it's hard to be safe:
```python
with torch.cuda.stream(s):
    # imagine forward used a lot of streams, so backward leaf nodes may run on many streams
    loss.backward()
    # backward() syncs the default stream with all the leaf streams, but does not sync s with anything,
    # so counterintuitively (even though we're in the same stream context as backward()!)
    # it is NOT SAFE to use grads here, and there's no easy way to make it safe,
    # unless you manually sync on all the streams you used in forward,
    # or move "use grads" back to default stream outside the context.
    use grads
```

This PR makes backward calls have the same stream semantics wrt surrounding ops as any other call. (https://github.com/pytorch/pytorch/pull/45787 did so for the relationship between root gradients and the beginning of backward. This PR tries to establish the same for the relationship between leaf gradients and ops following backward.)

See diffs in cuda.rst for the user-facing consequences.

[Comments added to Note [Streaming backwards]](https://github.com/pytorch/pytorch/pull/54227/files#diff-c66dfeac2a2da1867233047ec413c7e625644c672d7b38b8ec982f5605923c64R149-R167) explain the implementation.

This PR is relevant to CUDA graphs because graphs can't capture anything involving the default stream, so the default stream sync prevents capturing a full backward pass (see https://github.com/pytorch/pytorch/issues/15623#issuecomment-786865755, https://github.com/pytorch/pytorch/issues/15623#issuecomment-789362495) and prevents the "best case scenario" graphs use case
```python
with torch.cuda.stream(s):
    graph.capture_begin()
    loss = fwd(input)
    loss.backward()
    graph.capture_end()
```
This PR is the first step towards a graph-friendly situation where we _only_ sync the current stream, not the default stream, with the leaf streams. For temporary BC, this PR still syncs the default stream (in addition to the current stream) on the leaf streams. So it doesn't yet enable full-backward() capture, and keeps the weird BC pattern in the first code block above safe. Final BC-breaking removal of default-stream syncs will come in a followup PR.

### CODE REVIEW

**Comment by albanD** [comment]
I think the question related to different threads is linked to the fact that they are here to handle multiple GPUs.
In particular, shouldn't you be syncing with the current stream of all devices? What is the expected behavior there?

**Comment by mcarilli** [comment]
> In particular, shouldn't you be syncing with the current stream of all devices? What is the expected behavior there?

The way it's written now, if a leaf node ran on a particular device, we sync that device's current stream with that leaf stream, but we don't insert any cross-device syncs (ie, we don't ensure that current streams of all devices used by the leaf streams are synced against all the leaf stream).

File: docs/source/notes/cuda.rst [hunk 1]
```diff
@@ -201,20 +201,42 @@ ensure proper synchronization.
 Stream semantics of backward passes
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-Internally, each backward CUDA op runs on the same stream that was used for its corresponding forward op.
-
-When manually supplying CUDA tensor(s) as a backward pass's initial gradient(s) (e.g.,
+A. Each backward CUDA op runs on the same stream that was used for its corresponding forward op.
```
Comment (albanD) [inline docs/source/notes/cuda.rst:204]: Should we talk about device here?
The current device will be the device of the output of the op. If the first output of the op is not a cuda Tensor, the current device will be the same as the caller's one.

File: docs/source/notes/cuda.rst [hunk 2]
```diff
@@ -201,20 +201,42 @@ ensure proper synchronization.
...
+    # Unsafe
+    with torch.cuda.stream(strm):
+        loss.backward()
+    use grads
 
-have the same stream-semantics relationship as any pair of ops::
+    # Safe, with synchronization
+    with torch.cuda.stream(strm):
+        loss.backward()
+    torch.cuda.current_stream().wait_stream(strem)
```
Comment (albanD) [inline docs/source/notes/cuda.rst:236]: nit strem -> strm

File: torch/csrc/autograd/engine.cpp [hunk 3]
```diff
@@ -144,18 +144,30 @@ static thread_local std::shared_ptr<ReadyQueue> local_ready_queue = nullptr;
...
-// behavior the engine records "leaf streams," the streams of the
-// leaf variables, and syncs them with their device's default stream
-// at the end of backward. All other streams are already synchronized
-// to happen before at least one leaf stream (per the above), so syncing
-// the leaf streams with the default streams is sufficient to implement
-// the historic behavior.
+// The stream semantics of a backward() (or torch.autograd.grad()) call
+// with respect to surrounding ops are the same as for any other call:
+//
+// Safe
```
Comment (albanD) [inline torch/csrc/autograd/engine.cpp:150]: nit: maybe just point at the doc note on cuda? it is much more detailed

File: torch/csrc/autograd/engine.cpp [hunk 4]
```diff
@@ -880,6 +880,8 @@ auto Engine::execute(const edge_list& roots,
                      bool create_graph,
                      bool accumulate_grad,
                      const edge_list& outputs) -> variable_list {
+  const auto guard = c10::impl::VirtualGuardImpl{c10::DeviceType::CUDA};
+  // std::cout << "current stream " << guard.getStream(*at::device_of(inputs[0])) << std::endl;
```
Comment (albanD) [inline torch/csrc/autograd/engine.cpp:884]: nit debug print

File: torch/csrc/autograd/engine.cpp [hunk 5]
```diff
@@ -933,6 +935,21 @@ auto Engine::execute(const edge_list& roots,
   // in dist_engine.cpp).
   auto& fut = graph_task->future_result_;
   fut->wait();
+
+  // Syncs leaf streams with ambient (current) streams of the thread that called execute().
+  // See note "Streaming backwards". leaf_streams is an unordered_set so it should only
```
Comment (albanD) [inline torch/csrc/autograd/engine.cpp:940]: Which note do you refer to here? If it is in the website doc, you can add a link, otherwise we use [] to mark the note name and make it more grepable

File: torch/csrc/autograd/engine.cpp [hunk 6]
```diff
@@ -880,6 +880,8 @@ auto Engine::execute(const edge_list& roots,
                      bool create_graph,
                      bool accumulate_grad,
                      const edge_list& outputs) -> variable_list {
+  const auto guard = c10::impl::VirtualGuardImpl{c10::DeviceType::CUDA};
```
Comment (albanD) [inline torch/csrc/autograd/engine.cpp:883]: What is this guard doing? Just to capture the current stream at the beginning of the call?
Why not capture that at the end where we do the rest of the stream logic?

Ho you actually already do that there. So this is dead code?

**Review by albanD — COMMENTED** [review]
Should we update some tests?

File: torch/csrc/autograd/engine.cpp (hunk 3 above)
Comment (mcarilli) [reply torch/csrc/autograd/engine.cpp:150]: That's a better idea.

File: torch/csrc/autograd/engine.cpp (hunk 5 above)
Comment (mcarilli) [reply torch/csrc/autograd/engine.cpp:940]: Changed to Note [Streaming backward]. The Note gives a bit more low level description that isn't in the sphinx.

File: torch/csrc/autograd/engine.cpp (hunk 6 above)
Comment (mcarilli) [reply torch/csrc/autograd/engine.cpp:883]: It was just there to serve the debugging print below :P

File: docs/source/notes/cuda.rst (hunk 2 above)
Comment (mcarilli) [reply docs/source/notes/cuda.rst]: changed all "strm" to "s" to match other examples on the page.

File: test/test_cuda.py [hunk 7]
```diff
@@ -1729,6 +1736,62 @@ def accum_hook(grad):
 
         self.assertEqual(x.grad, torch.ones_like(x) * 5)
 
+    # Skip the test for ROCm as per https://github.com/pytorch/pytorch/issues/53190
+    @skipIfRocm
```
Comment (mruberry) [inline test/test_cuda.py:1740]: What about a simple test where a new stream is used, an operation is performed, backward happens, and the grads are checked on that same stream?

```
with stream s0:
  do some ops on a tensor
  backward
  check the grad and it's correct
```

**Review by mruberry — APPROVED** [review]
Aside from a testing recommendation this now LGTM. @albanD, would you like to take another look?

Also @mcarilli, would you edit the PR title and description to reflect that this is no longer a BC-breaking change, and clarify what this PR is actually doing now.

File: torch/csrc/autograd/engine.cpp [hunk 8]
```diff
@@ -898,6 +924,11 @@ auto Engine::execute(const edge_list& roots,
       /* depth */ not_reentrant_backward_call ? 0 : total_depth + 1,
       /* cpu_ready_queue */ local_ready_queue);
 
+  // Collects current and default streams for devices where this process has a context,
+  // so GraphTask::exec_post_processing can sync them with leaf_streams.
+  // Should this be inside init_to_execute?
```
Comment (albanD) [inline torch/csrc/autograd/engine.cpp:929]: no, that function (unlike what the name suggests) is not always called and is used for very specific cases only. Here is good.

File: torch/csrc/autograd/engine.h [hunk 9]
```diff
@@ -112,6 +112,14 @@ struct GraphTask: std::enable_shared_from_this<GraphTask> {
 
   std::unordered_set<c10::Stream> leaf_streams;
 
+  // Per-device current and default streams of the execute() that called this GraphTask.
+  // These will be synced with leaf_streams in exec_post_processing.
+  std::vector<c10::optional<c10::Stream>> caller_current_streams_;
+  std::vector<c10::optional<c10::Stream>> caller_default_streams_;
+
+  // Collects caller_current_streams_ and caller_default_streams_
+  void stash_current_streams();
```
Comment (albanD) [inline torch/csrc/autograd/engine.h:121]: cc @pritamdamania  should we add a call to this in the distributed engine to make sure it benefits from this proper synchronization as well?

**Review by albanD — APPROVED** [review]
LGTM!
Thanks for the update.

We might need to update the distributed engine to behave well as well, but that can be a follow up PR.

**Comment by mcarilli** [comment]
> would you edit the PR title and description to reflect that this is no longer a BC-breaking change, and clarify what this PR is actually doing now.

@mruberry Updated title and description, hopefully more accurate now.

File: test/test_cuda.py (hunk 7 above)
Comment (mcarilli) [reply test/test_cuda.py:1757]: Added slightly more challenging version
```python
with s0:
    do some ops on a tensor
with s1:
    sync s1 with s0
    backward
    check the grad and it's correct
```
to [test_streaming_backwards_sync](https://github.com/pytorch/pytorch/pull/54227/files#diff-893b1eea27352f336f4cd832919e48d721e4e90186e63400b8596db6b82e7450R1699-R1713) above.

File: torch/csrc/autograd/engine.h (hunk 9 above)
Comment (pritamdamania87) [reply torch/csrc/autograd/engine.h:121]: Yes, I think this makes sense.

File: torch/csrc/autograd/engine.cpp [hunk 10]
```diff
@@ -864,6 +906,9 @@ auto Engine::compute_dependencies(Node* root, GraphTask& task, uint64_t min_topo
     if (fn->topological_nr() < min_topo_nr) {
       continue;
     }
+    if (!will_use_cuda) {
+      will_use_cuda = fn->stream(c10::DeviceType::CUDA).has_value();
+    }
```
Comment (mcarilli) [inline torch/csrc/autograd/engine.cpp:911]: @albanD I'm piggybacking on `compute_dependencies` for an advance check if any backward node (including leaf nodes) will use cuda, to decide [if i I need to bother remembering the current streams](https://github.com/pytorch/pytorch/pull/54227/files#diff-c66dfeac2a2da1867233047ec413c7e625644c672d7b38b8ec982f5605923c64R957-R961). Is this the right spot to check if each node has a CUDA stream()? Should i move the check a few lines up, above
```C++
    if (fn->topological_nr() < min_topo_nr) {
      continue;
    }
```?
